File: packages/backend/apps/translations/management/commands/export_translations.py
"""
Export translations from the database to JSON files.

This command exports all active locale translations to JSON files
that can be used by the frontend application.

Usage:
    python manage.py export_translations [--output-dir PATH]
"""

import json
from pathlib import Path

from django.core.management.base import BaseCommand

from apps.translations.models import Locale, Translation, TranslationKey


class Command(BaseCommand):
    help = 'Export translations from database to JSON files'

    def add_arguments(self, parser):
        parser.add_argument(
            '--output-dir',
            type=str,
            help='Output directory for JSON files (defaults to /app/translations)',
            default='/app/translations',
        )
        parser.add_argument(
            '--format',
            type=str,
            choices=['full', 'simple', 'json-dump'],
            default='full',
            help='Output format: full (with defaultMessage wrapper), simple (key-value), json-dump (all data as JSON to stdout)',
        )

    def handle(self, *args, **options):
        output_dir = Path(options['output_dir'])
        output_format = options['format']

        # If json-dump format, output everything to stdout and exit
        if output_format == 'json-dump':
            self.dump_all_translations_as_json()
            return

        # Ensure output directory exists
        output_dir.mkdir(parents=True, exist_ok=True)

        # Get master translations (for fallback)
        master_keys = {key.key: key.default_message for key in TranslationKey.objects.filter(is_deprecated=False)}
        total_keys = len(master_keys)

        self.stdout.write(f"Found {total_keys} active translation keys")
        self.stdout.write("")

        # Get all active locales
        locales = Locale.objects.filter(is_active=True).order_by('sort_order')

        locales_info = []

        for locale in locales:
            # Get all published translations for this locale
            translations = Translation.objects.filter(
                locale=locale, status=Translation.Status.PUBLISHED, key__is_deprecated=False
            ).select_related('key')

            # Build translation dict
            locale_translations = {t.key.key: t.value for t in translations}
            translated_count = len(locale_translations)

            # Build output based on format
            if output_format == 'full':
                # Full format with defaultMessage wrapper
                output_data = {}
                for key, default_msg in master_keys.items():
                    output_data[key] = {'defaultMessage': locale_translations.get(key, default_msg)}
            else:
                # Simple key-value format
                output_data = {}
                for key, default_msg in master_keys.items():
                    output_data[key] = locale_translations.get(key, default_msg)

            # Write to file
            output_path = output_dir / f"{locale.code}.json"
            with open(output_path, 'w', encoding='utf-8') as f:
                json.dump(output_data, f, ensure_ascii=False, indent=2)

            # Calculate percentage
            percent = (translated_count / total_keys * 100) if total_keys > 0 else 0
            rtl_marker = ' (RTL)' if locale.rtl else ''

            self.stdout.write(
                self.style.SUCCESS(
                    f"  ✓ {locale.code}.json - {translated_count}/{total_keys} keys ({percent:.1f}%){rtl_marker}"
                )
            )

            locales_info.append(
                {
                    'code': locale.code,
                    'name': locale.name,
                    'native_name': locale.native_name,
                    'rtl': locale.rtl,
                    'is_default': locale.is_default,
                }
            )

        # Write locales.json metadata file
        locales_path = output_dir / "locales.json"
        with open(locales_path, 'w', encoding='utf-8') as f:
            json.dump(locales_info, f, ensure_ascii=False, indent=2)

        self.stdout.write(self.style.SUCCESS(f"  ✓ locales.json - {len(locales_info)} locales"))
        self.stdout.write("")
        self.stdout.write(self.style.SUCCESS(f"Export complete! Files written to: {output_dir}"))

    def dump_all_translations_as_json(self):
        """Dump all translations as JSON to stdout for external processing."""
        result = {'locales': [], 'translations': {}}

        # Get master keys
        master_keys = {key.key: key.default_message for key in TranslationKey.objects.filter(is_deprecated=False)}

        for locale in Locale.objects.filter(is_active=True).order_by('sort_order'):
            result['locales'].append(
                {
                    'code': locale.code,
                    'name': locale.name,
                    'native_name': locale.native_name,
                    'rtl': locale.rtl,
                    'is_default': locale.is_default,
                }
            )

            translations = Translation.objects.filter(
                locale=locale, status=Translation.Status.PUBLISHED, key__is_deprecated=False
            ).select_related('key')

            locale_translations = {t.key.key: t.value for t in translations}

            # Include all keys with fallback to master
            full_translations = {}
            for key, default_msg in master_keys.items():
                full_translations[key] = locale_translations.get(key, default_msg)

            result['translations'][locale.code] = full_translations

        # Output as single-line JSON to stdout
        self.stdout.write(json.dumps(result, ensure_ascii=False))

File: packages/webapp/scripts/exportTranslations.js
#!/usr/bin/env node
/**
 * Export Translations Script
 *
 * This script exports translations from the backend database to JSON files
 * in the translations directory. Each active locale gets its own JSON file.
 *
 * Usage:
 *   node scripts/exportTranslations.js
 *
 * Output:
 *   packages/webapp-libs/webapp-core/src/translations/{locale}.json
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

const TRANSLATIONS_DIR = path.join(
  __dirname,
  '../../webapp-libs/webapp-core/src/translations'
);

/**
 * Check if docker compose backend container is running
 */
function isBackendContainerRunning() {
  try {
    const result = execSync('docker compose ps --format json backend', {
      encoding: 'utf8',
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    const lines = result.trim().split('\n').filter(Boolean);
    return lines.some((line) => {
      try {
        const container = JSON.parse(line);
        return container.State === 'running';
      } catch {
        return false;
      }
    });
  } catch {
    return false;
  }
}

/**
 * Export translations from the database
 */
async function exportTranslations() {
  console.log('📤 Translation Export Script');
  console.log('='.repeat(50));

  if (!isBackendContainerRunning()) {
    console.error('❌ Backend container is not running.');
    console.error('');
    console.error('Start the backend with: pnpm saas backend up');
    process.exit(1);
  }

  console.log('🔍 Fetching translations from database...');

  try {
    // Use the Django management command with json-dump format
    const output = execSync(
      'docker compose exec -T backend python manage.py export_translations --format=json-dump',
      {
        encoding: 'utf8',
        maxBuffer: 50 * 1024 * 1024, // 50MB buffer for large translation data
        stdio: ['pipe', 'pipe', 'pipe'],
      }
    );

    // Parse the JSON output (skip any warning lines)
    const lines = output.split('\n');
    let jsonLine = '';
    for (const line of lines) {
      if (line.startsWith('{')) {
        jsonLine = line;
        break;
      }
    }

    if (!jsonLine) {
      throw new Error('No JSON output from database export');
    }

    const data = JSON.parse(jsonLine);

    console.log(`📚 Found ${data.locales.length} active locales`);
    console.log('');

    // Ensure translations directory exists
    if (!fs.existsSync(TRANSLATIONS_DIR)) {
      fs.mkdirSync(TRANSLATIONS_DIR, { recursive: true });
    }

    // Get total keys count from first locale
    const sampleLocale = data.locales[0]?.code;
    const totalKeys = sampleLocale ? Object.keys(data.translations[sampleLocale] || {}).length : 0;

    // Export each locale
    for (const locale of data.locales) {
      const localeCode = locale.code;
      const translations = data.translations[localeCode] || {};
      const translationCount = Object.keys(translations).length;

      // Build output with defaultMessage wrapper
      const outputData = {};
      for (const [key, value] of Object.entries(translations)) {
        outputData[key] = { defaultMessage: value };
      }

      const outputPath = path.join(TRANSLATIONS_DIR, `${localeCode}.json`);
      fs.writeFileSync(
        outputPath,
        JSON.stringify(outputData, null, 2),
        'utf8'
      );

      const percent = totalKeys > 0 ? ((translationCount / totalKeys) * 100).toFixed(1) : 0;
      const rtlMarker = locale.rtl ? ' (RTL)' : '';

      console.log(`  ✅ ${localeCode}.json - ${translationCount}/${totalKeys} keys (${percent}%)${rtlMarker}`);
    }

    // Also create a locales.json file with metadata
    const localesPath = path.join(TRANSLATIONS_DIR, 'locales.json');
    fs.writeFileSync(
      localesPath,
      JSON.stringify(data.locales, null, 2),
      'utf8'
    );
    console.log(`  ✅ locales.json - ${data.locales.length} locales`);

    // Create translations_export.json for backend deployment
    // This file is used by the backend to import translations during deployment
    const exportData = {
      locales: data.locales,
      translations: data.translations,
    };
    const exportPath = path.join(TRANSLATIONS_DIR, 'translations_export.json');
    fs.writeFileSync(
      exportPath,
      JSON.stringify(exportData, null, 2),
      'utf8'
    );
    console.log(`  ✅ translations_export.json - deployment bundle`);

    console.log('');
    console.log('✅ Export complete!');
    console.log(`   Output directory: ${TRANSLATIONS_DIR}`);

  } catch (error) {
    console.error('❌ Export failed:', error.message);
    process.exit(1);
  }
}

// Run the export
exportTranslations().catch(console.error);
